ddl/ddl_api.go

```diff
@@ -358,7 +385,16 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType, tblCharset string, dbCh
 // outPriKeyConstraint is the primary key constraint out of column definition. such as: create table t1 (id int , age int, primary key(id));
 func buildColumnAndConstraint(ctx sessionctx.Context, offset int,
 	colDef *ast.ColumnDef, outPriKeyConstraint *ast.Constraint, tblCharset, dbCharset string) (*table.Column, []*ast.Constraint, error) {
-	if err := setCharsetCollationFlenDecimal(colDef.Tp, tblCharset, dbCharset); err != nil {
+	// The specified collate is in colDef.Options, but the charset is in colDef.Tp.
+	specifiedCollate := ""
+	for _, op := range colDef.Options {
```

Comment:
The colDef.options is iterated in the https://github.com/pingcap/tidb/blob/master/ddl/ddl_api.go#L2062, Could we eliminate the iteration here?

Reply:
I've checked it!  
The `setCharsetCollationFlenDecimal` func is called by two different logic: one is create table, the other is modify column. 
For the later there is an option loop indeed before call this func, if we try to make use of loop, we need pass parameter, it's ok. Consequently, for the former that's means we have no option loop neither in the func or before call it, it will cause an error.
So we still need the loop here.

Reply:
> The colDef.options is iterated in the https://github.com/pingcap/tidb/blob/master/ddl/ddl_api.go#L2062, Could we eliminate the iteration here?

I've checked it!
The setCharsetCollationFlenDecimal func is called by two different logic: one is create table, the other is modify column.
For the later there is an option loop indeed before call this func, if we try to make use of loop, we need pass parameter, it's ok. Consequently, for the former that's means we have no option loop neither in the func or before call it, it will cause an error.
So we still need the loop here.